### ISSUE TITLE
 nushell Startup Time became extremely slow since 0.94.0

### ISSUE THREAD

**Issue by FantasqueX** [nushell/nushell#13023]
### Describe the bug

Before 0.94.0, such as 0.93.0, nushell's startup time is super-fast, normally 20ms~30ms. However, since 0.94.0, it becomes extremely slow, more than 2min. My environment is ArchLinux in WSL. Both ArchLinux build and official is affected.

### How to reproduce

1. Download 0.94.1 official release tarball
2. extract
3. execute `./nu`
4. Startup Time: 2min 9sec 940ms 732µs 995ns

### Expected behavior

Startup time is around 30ms.

### Screenshots

![image](https://github.com/nushell/nushell/assets/46747123/9c0ea3e8-0904-4fe4-9a15-810a6464f80e)

### Configuration

| key                | value                                                                |
| ------------------ | -------------------------------------------------------------------- |
| version            | 0.94.1                                                               |
| major              | 0                                                                    |
| minor              | 94                                                                   |
| patch              | 1                                                                    |
| branch             |                                                                      |
| commit_hash        | f3cf693ec701ab3086f984a8aeadfc9b50fadd90                             |
| build_os           | linux-x86_64                                                         |
| build_target       | x86_64-unknown-linux-gnu                                             |
| rust_version       | rustc 1.77.2 (25ef9e3d8 2024-04-09)                                  |
| rust_channel       | 1.77.2-x86_64-unknown-linux-gnu                                      |
| cargo_version      | cargo 1.77.2 (e52e36006 2024-03-26)                                  |
| build_time         | 2024-05-30 23:10:14 +00:00                                           |
| build_rust_channel | release                                                              |
| allocator          | mimalloc                                                             |
| features           | default, sqlite, static-link-openssl, system-clipboard, trash, which |
| installed_plugins  |                                                                      |

**Comment by fdncred** [comment]
Very interesting. You're the first person that has reported a slow down in startup time. Most people reported a significantly faster startup time when we moved to a msgpackz plugin repository.

**Comment by FantasqueX** [comment]
Hi, thanks for your advice. I have tried it. Here is the result.
![image](https://github.com/nushell/nushell/assets/46747123/1229f954-d40a-45dd-a48c-b3d6af3d1726)
It looks like that my config is not the root cause at least for ArchLinux build.

**Comment by fdncred** [comment]
I'm not sure how to help since I can't reproduce the problem.

**Comment by s-cerevisiae** [comment]
Since you are using WSL, maybe you can try to [remove the Windows' `PATH` from environment](https://stackoverflow.com/questions/51336147/how-to-remove-the-win10s-path-from-wsl) as that can slow everything down significantly.

If that works maybe nushell is doing something different with `PATH` in the new version.

**Comment by FantasqueX** [comment]
> Since you are using WSL, maybe you can try to [remove the Windows' `PATH` from environment](https://stackoverflow.com/questions/51336147/how-to-remove-the-win10s-path-from-wsl) as that can slow everything down significantly.
> 
> If that works maybe nushell is doing something different with `PATH` in the new version.

Thanks for your advice. I have tried this method. Unfortunately, the startup time is still slow :(

**Comment by fdncred** [comment]
Are you trying to us any plugins? I notice that your version doesn't show any but I wanted to ask.

**Comment by FantasqueX** [comment]
> Are you trying to us any plugins? I notice that your version doesn't show any but I wanted to ask.

No.

**Comment by sadguitarius** [comment]
I'm getting this too. Only on one of my two machines running WSL. I'm also running Arch on both machines, so that could be something there. I don't have it happening with the native Windows version. I can't find any meaningful difference between the two setups yet. Dotfiles are synced, and environment variables are mostly the same. I did try removing the Windows path from the environment and it made no difference, as well as starting nu from a clean config, which also didn't fix anything.

**Comment by fdncred** [comment]
WSL historically is slow if you don't take steps to make it faster. Those steps are mainly with how you setup WSL in regards to Windows paths. For instance, on my WSL, I have all linux entries in my path and 2 windows paths. However, if I let WSL do what it wants, I'd have 100 Windows path entries and 10 linux entries which would bring WSL to a crawl.

**Comment by rogtino** [comment]
I'm also facing the same issue.
Here is my config which might be useful to repro:
.wslconfig
```
[experimental]
autoMemoryReclaim=gradual  # gradual  | dropcache | disabled
networkingMode=mirrored
dnsTunneling=true
firewall=true
```
/etc/wsl.conf
```
[boot]
systemd=true

[automount]
enabled = true
options = "metadata"
mountFsTab = true
[interop]
appendWindowsPath = false

```

**Comment by sadguitarius** [comment]
Just tried your config on my laptop which hasn't been having the issue and nothing bad happened. I'm going to be traveling for the next couple of weeks so won't have access to my desktop which is the one having the problem.

**Comment by sadguitarius** [comment]
Ah that must be it, I had just added that to my config on my desktop pc. No idea why that would be causing the problem though

**Comment by sadguitarius** [comment]
Just in case it helps, I enabled that option so I could use distcc for remote compilation and it's running a bunch of background processes. Unfortunately I won't be able to try anything on that computer for a while but that's the only other thing I can think of that's different between my 2 setups.

**Comment by fdncred** [comment]
Thanks for the updates and the resolution. Closing this now.

**Comment by FantasqueX** [comment]
@fdncred Why is this issue closed? I can confirm nushell's starup time become fast when using NAT networking mode. However, mirrored networking mode is very convenient. Why can't nushell and mirrored networking mode co-exist?

**Comment by fdncred** [comment]
I closed it because it doesn't seem to be a problem with nushell but a problem with WSL. Seems like mirrored networking was just introduced so I figured it will have some bugs. Nushell doesn't really know about networking at all, so I don't see how any networking issue like this could be a problem with nushell. Please feel free to prove me wrong and point me to the code in nushell where the problem is.

**Comment by FantasqueX** [comment]
Maybe I can bisect between 0.93.0 and 0.94.0 :)

**Comment by sadguitarius** [comment]
Yeah I think it's kind of odd to close the issue when as far as I can tell, using any other shell works fine with no lag. With Nushell being a shell that prides itself on cross-platform compatibility (which is one of the reasons I started using it in the first place), maybe there's a better approach? If a recent change to Nushell, not WSL, caused this issue to flare up, it would be good to get to the bottom of it. And maybe leave the issue open until the source of it is found.

**Comment by FantasqueX** [comment]
After bisecting, I found out that the culprit  is https://github.com/nushell/nushell/commit/3143ded374668fc6f404fc81044c680c06cb823d
However, I don't know how to debug next.

**Comment by FantasqueX** [comment]
According to https://github.com/nushell/nushell/pull/12469#issuecomment-2094858595
I noticed that if I don't update other deps, only `cargo remove divan --dev` and `cargo add tango-bench --dev`, nushell works well. Maybe some dependencies updates cause the issue.

**Comment by Kissaki** [comment]
I noticed significant slower without WSL, on Windows natively, but have not come around to testing it yet. After updating my configs from several version ago to current state it took multiple seconds to launch and prompt.

I will report back when I can test and narrow down. *(For now, just a note that this may be a general issue, not only on WSL.)*

**Comment by rgwood** [comment]
Thank you for bisecting+investigating this!

cc: @FilipAndersson245 who implemented that PR.

Skimming the list of updated dependencies, arboard is a little suspicious; that’s the first thing I’d try downgrading.

**Comment by FilipAndersson245** [comment]
Interesting, the code addition itself should not affect non test builds, so I guess it must be some `lock` dependency. Is there a simple way to debug what dependency did this?

**Comment by j-xella** [comment]
Please don't close this issue. It does happening, and not just with WSL. You can find more details in #13019 , but short summary here:

* Happens on RedHat Linux server for me - not WSL specific
* No such problems with 0.93
* it seems to be intermittent in my case - one day there is this problem, another day there is not. I am on a corporate server and not an admin, so can not control what else is going on there.
* Not config specific - `nu -n` has the same problem
* One-liners like `nu -c "commands"` do not seem to be slow, it is only the interactive shell that takes time to start.
* While starting, nu does not seem to be consuming a lot of CPU. So it is probably just waiting for something...
* Running `nu --log-level trace`, last message before the wait (nu 0.94.2 linux musl build) is 
```
2024-06-17 11:07:00.405 AM [INFO ] nu_utils::utils: perf: crates/nu-cli/src/repl.rs:96:9 translate env vars took 1.114557ms
```

**Comment by fdncred** [comment]
> One-liners like nu -c "commands"

This would seem to indicate that it has something to do with your env.nu, config.nu, or login.nu since none of them are loaded with `nu -c`.

**Comment by rlonstein** [comment]
This is happening with 0.95. I can't cut and paste since it's on a corporate host running RHEL.

Tracing without config and stdlib shows 9.3ms spent creating menus and 14.2ms in sync_history.

**Comment by sadguitarius** [comment]
I was also able to narrow this down in my case to the use of WSL mirrored mode networking. When the option was turned on in .wslconfig, Nushell would hang for about 2 minutes. With mirrored mode networking disabled, Nushell loads normally.

Running `nu --log-level trace` with mirrored mode enabled showed Nushell hanging in between the lines
`2024-07-24 08:40:17.829 PM [INFO ] nu_cli::repl: perf: crates/nu-cli/src/repl.rs:92:5 translate env vars took 926.9µs`
and
`2024-07-24 08:42:28.301 PM [INFO ] nu_cli::repl: perf: crates/nu-cli/src/repl.rs:221:5 setup reedline took 130.471836005s`

It sounds like issue is not necessarily only related to the use of WSL, but maybe this could help narrow things down. Again, I don't experience this problem using other shells in WSL, and I still have the problem running Nushell with a clean config and environment.
